mart/attack/projector.py

```diff
@@ -4,42 +4,58 @@
 # SPDX-License-Identifier: BSD-3-Clause
 #
 
-import abc
-from typing import Any, Dict, List, Optional, Union
+from __future__ import annotations
 
-import torch
+from typing import Any
 
-__all__ = ["Projector"]
+import torch
 
 
-class Projector(abc.ABC):
+class Projector:
     """A projector modifies nn.Parameter's data."""
 
     @torch.no_grad()
     def __call__(
         self,
-        tensor: torch.Tensor,
-        input: torch.Tensor,
-        target: Union[torch.Tensor, Dict[str, Any]],
+        perturbation: torch.Tensor | tuple,
+        *,
+        input: torch.Tensor | tuple,
+        target: torch.Tensor | dict[str, Any] | tuple,
+        **kwargs,
+    ) -> None:
+        if isinstance(perturbation, tuple):
+            for perturbation_i, input_i, target_i in zip(perturbation, input, target):
+                self.project(perturbation_i, input=input_i, target=target_i)
+        else:
+            self.project(perturbation, input=input, target=target)
+
+    def project(
+        self,
+        perturbation: torch.Tensor | tuple,
+        *,
+        input: torch.Tensor | tuple,
+        target: torch.Tensor | dict[str, Any] | tuple,
     ) -> None:
         pass
 
 
 class Compose(Projector):
     """Apply a list of perturbation modifier."""
 
-    def __init__(self, projectors: List[Projector]):
+    def __init__(self, projectors: list[Projector]):
```

Comment:
Do you think it's better to make it a dictionary so that it's more easily configurable in Hydra? Like that in Enforcer: https://github.com/IntelLabs/MART/blob/70e62d901f2a27327dca517ed21a12d81dd49040/mart/attack/enforcer.py#L100

The keys in the Enforcer constraints dictionary are not used. Maybe we can use that for modality configuration?

Reply:
I think that should live in another PR.

Reply:
My read of the Projectors class is that they need a wholesale redesign because there's a lot of mixing of functionality instead of focusing on composition.